File: python/MQReceiver.py
#!/usr/bin/env python
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
# 
#   http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

import sys
from qpid.messaging import *

if __name__ == "__main__":
    if len(sys.argv) < 5:
        sys.exit(-1)
    print 'app name {}, broke ip {}, broker port {}, queue id {}'.format(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4])
    broker = "{}/xxxx@{}:{}".format(sys.argv[1], sys.argv[2], sys.argv[3])
    address = "{}".format(sys.argv[4])
    conn_options = {
                    'transport'               : 'ssl',
                    'ssl_keyfile'             : "ssl_cert_file/MSP.Key.pem",
                    'ssl_certfile'            : "ssl_cert_file/MSP.pem.cer",
                    'ssl_trustfile'           : "ssl_cert_file/Wireless Root CA.pem.cer",
                    'ssl_skip_hostname_check' : True,
                    }
    connection = Connection(broker, **conn_options)

    try:
        connection.open()
        session = connection.session()
        receiver = session.receiver(address)
        print "session create success"
        while True:
            message = receiver.fetch()
            if len(sys.argv) == 6:
                if message.content.find(sys.argv[5]):
                    print "%r" % message.content
            else:
                print "%r" % message.content
            session.acknowledge()

    except MessagingError, m:
        print "MessagingError", m
    connection.close()

File: cpp/MQReceiver.cpp
/*******************************************
This is an example of receive AMQP message
author:  renyu@huawei.com
date: 2015-03-10
version: v0.1
*******************************************/

#include <stdlib.h>
#include <iostream>
#include <string>
#include <qpid/messaging/Connection.h>
#include <qpid/messaging/Message.h>
#include <qpid/messaging/Receiver.h>
#include <qpid/messaging/Sender.h>
#include <qpid/messaging/Session.h>

using namespace qpid::messaging;

int main(int argc, char** argv)
{
    if ( argc != 5 )
    {
        std::cout << "Usage: MQReceiver app_name ip port queue_id" << std::endl;
        return 0;
    }
    
    std::string app_name = argv[1];
    std::string ip       = argv[2];
    std::string port     = argv[3];
    std::string queue_id = argv[4];    
    
    std::string address      = queue_id;    
    std::string broker       = app_name + "/xxxx@" + ip + ":" + port;    
    
    try {
        setenv("QPID_SSL_CERT_DB", "cert_db", 1);

        Connection connection(broker);

        connection.setOption("sasl-mechanism", "PLAIN");
        connection.setOption("username", "msp_mq_access");
        connection.setOption("password", "xxxx");

        connection.setOption("ssl-cert-name", "msp");
        connection.setOption("transport", "ssl");

        connection.open();
        Session session = connection.createSession();
        
        Receiver receiver = session.createReceiver(address);
        receiver.setCapacity(1024);
        
        while(true)
        {
            Message message = receiver.fetch();            
            std::cout << message.getContent() << std::endl;            
            session.acknowledge();
        }
        
    } catch(const std::exception& error) {
        std::cerr << error.what() << std::endl;
        return 1;   
    }
}
